### ISSUE TITLE
Rule proposal: only allow function calls starting with underscore on `this`

### ISSUE THREAD

**Issue by janpaul123** [eslint/eslint#3435]
We have the convention to prefix private functions with an underscore. It would be useful to disallow direct calls to private functions unless from the same object, which can be reasonably be approximated by implementing a rule like this:

Good:
```javascript
  someObject.publicFunction();
  getSomeObject().publicFunction();
  someCallback();
  _someCallbackThatHappensToStartWithAnUnderscore();
  this._privateFunction();
```

Bad:
```javascript
  someObject._privateFunction();
  getSomeObject()._privateFunction();
```

To answer the questions as suggested:
- The use case for the rule - what is it trying to prevent or flag? Include code examples.

This is to prevent using "private functions" (as commonly prefixed by an underscore) from outside the class that has "access" to them. This makes changing the internals of a class easier, since you have a stronger guarantee that private functions are not called. It thus prevents errors when changing the internals of a class.

- Whether the rule is trying to prevent an error or is purely stylistic.

It would prevent an error, to some degree. Or to be more precise, it gives stronger guarantees that make errors less likely.

- Why you believe this rule is generic enough to be included.

Yes, as far as I know prefixing private methods with an underscore is common practice.

- Whether the rule should be on or off by default.

No, too high chance of false positives.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by gyandeeps** [comment]
Based on our docs, these are the points we touch on when deciding on new rules.
Link: http://eslint.org/docs/developer-guide/contributing#new-rules

Can you please provide more info.
We also provide the ability to create project or company specific rules by using plugin architecture. These come in handy specially when you have a usecase which is not generic enough.
More info: http://eslint.org/docs/developer-guide/working-with-plugins

**Comment by janpaul123** [comment]
@gyandeeps I updated the original post with the points you asked about. I think this could benefit the wider developer community. Please let me know if it makes sense!

**Comment by mysticatea** [comment]
Sounds good to me.

This seems to be gotten with making the `no-underscore-dangle` rule ignoring next of `this.`, but the name `no-underscore-dangle` is not expressing this purpose.

Hmm...

My idea, more generalized:

```json
{
    "no-member-access-violation": [2, {"private": "^_", "protected": "^_"}]
}
```

- `private` is a regular expression for the name of private members.
  This rule warns `Identifier`s (excludes next of `this.`) which matches the regular expressions.
- `protected` is a regular expression for the name of protected members.
  This rule warns `Identifier`s (excludes next of `this.` or `super.`) which matches the regular expressions.

Note: `protected` is enabled in ES6 Classes.

**Comment by nzakas** [comment]
If there is a change to be made, it would have to be for `no-underscore-dangle`, otherwise we'd have rules with overlapping responsibilities.

**Comment by mysticatea** [comment]
I thought my proposal is not related with underscore directly. But I don't oppose that the general rule is not needed.

**Comment by just-boris** [comment]
Sent pull-request #3822 to resolve it

**Comment by nzakas** [comment]
I don't think we have identified the best solution yet, so let's keep talking.

Ultimately, I think we want to say "allow dangling underscore if preceded by `this`". Does everyone agree with that?

**Comment by mysticatea** [comment]
Agreed

**Comment by just-boris** [comment]
@nzakas @mysticatea, so you want to keep this rule without configuration, but add exception regarding `this` keyword?

**Comment by nzakas** [comment]
Yes, that's the proposal. @janpaul123 would that work for you?

**Comment by just-boris** [comment]
Updated my PR. Now it consists only exception for properties after `this`

**Comment by janpaul123** [comment]
Thanks all so much, this is a perfect solution! :+1:

**Comment by nzakas** [comment]
@just-boris thanks, the PR looks good.

I just want to get a final tally from everyone to make sure this is the best solution. The current solution is to just ignore (automatically, for everyone) properties that are preceded by `this`, so `this._foo()` now no longer warns for anyone. Is that the right solution? (Other approach: add an option to turn this behavior on instead of doing it by default.)

@eslint/eslint-team thoughts?

**Comment by janpaul123** [comment]
Ping, it has been a few weeks. Would love to see this get merged! :)

**Comment by IanVS** [comment]
@nzakas, I missed this before, but I think that principle of least surprise dictates that this should be an option, not a default.

**Comment by ilyavolodin** [comment]
I think I would go with an option too.

**Comment by nzakas** [comment]
@janpaul123 there's no open PR for this. 

Okay, let's go for an option. Suggestions for a name?

**Comment by ilyavolodin** [comment]
`excludePrivate`?

**Comment by IanVS** [comment]
`exceptThis`?

**Comment by janpaul123** [comment]
@nzakas I was referring to https://github.com/eslint/eslint/pull/3822, but an option for this seems reasonable!

**Comment by just-boris** [comment]
Well, this is what I did at first, but then @nzakas asked me to rewrite it without option.
:disappointed:

**Comment by nzakas** [comment]
@janpaul123 yup, that PR is closed.

@just-boris I'm sorry, sometimes that happens when discussing solutions. That's why we recommend people wait until the discussion concludes before starting to implement something. We mark issues with the "accepted" label when they are ready for implementation.

**Comment by just-boris** [comment]
Well, now issue has `accepted` label. And what's the final conclusion about property name?
It matters for me, because I already using my fork and I want to switch back to master as it will be possible

**Comment by nzakas** [comment]
We are still trying to determine the option name (see last couple comments).

I think I like `allowThis` the best. What does everyone else think?

**Comment by just-boris** [comment]
As for me, `true` value of the option should enable dangles after `this`. So, `allowThis` is ok.

**Comment by ilyavolodin** [comment]
`allowThis` sounds good to me.

**Comment by nzakas** [comment]
OK, let's do it

**Comment by just-boris** [comment]
Just found that you already have configuration option named `allow`. It is whitelist of identifiers which allowed to be with dangles. `allowThis` in this context means that it allows `this` identifier and it is confusing.

So, what about to name new option `allowAfterThis`? It will resolve the conflict in options meanings

**Comment by just-boris** [comment]
I almost finished implementation, but stuck with naming issue

**Comment by nzakas** [comment]
Sounds good.

**Comment by just-boris** [comment]
Done. PR #4562
